### DISCUSSION TITLE
0.23 Serving binary content

### DISCUSSION THREAD

**Discussion by petrm** [pocketbase/pocketbase#5940] [Q&A] [answered]
Hi,
I used to serve binary content with:
```
return e.blob(200, "application/octet-stream", record.get("content"));
```

This is now failing with an exception:
```
pocketbase  | ERROR GET /boot/x86_64/openssl-3.3.2-r1.apk
pocketbase  | └─ Something went wrong while processing your request.
pocketbase  |    └─ [PANIC RECOVER] template: err:12: unrecognized character in action: U+FFFD '�' goroutine 100 [running]:
```

I don't see any blob method in the docs anymore. What can I use instead?

**Comment by ganigeorgiev** [comment, marked as answer]
There is no `Blob` helper at the moment (_~I'll consider adding it in the next minor release~ Added in the `develop` branch_).

For now you can try writing the response manually with something like:

```js
e.response.header().set("content-type", "application/octet-stream")
e.response.writeHeader(200)
e.response.write(yourData)
return null // optional
```

_(side-note: while there is no `Blob` helper, note that the panic indicates that the error is maybe caused by something else)_

**Reply by petrm** [reply]
Many thanks, this solved the blob issue.
However, I am still getting that exception and it seems to be related to persisting binary content. It seems to be triggered by `$app.save(record);`. This worked fine with 0.22.

**Reply by ganigeorgiev** [reply]
Without a more complete code sample I'm not sure how to help.

The panic indicates that maybe there is a template error somewhere (e.g. if you have a pre/post save hook that sends email).

**Comment by petrm** [comment]
I replaced `$app.save(record);` with `$app.saveNoValidate(record);` and it works now. My table with files does not have a mime type set.

**Reply by ganigeorgiev** [reply]
I don't recommend skipping the validations as it could hide a bug.

I cannot read your mind, nor I have any context about your project to understand what is causing it to fail. Try to provide a more complete code sample that can reproduce the problem.

Note that you can also check the _Dashboard > Logs_  to see if there are more details related to the panic.

**Reply by petrm** [reply]
Something to support the mind reading exercise :)


Here is the minimal example. The table has only two fields, file(content) and string(filename).

```
routerAdd("GET", "/boot/x86_64/{filename}", (e) => {
    const util = require(`${__hooks}/util.js`);
    const filename = e.request.pathValue("filename");
    try {
        const record = $app.findFirstRecordByData("apk", "filename", filename);
        console.log("found file " + filename);
        return util.blob(e, 200, "application/octet-stream", record.get("content"));
    } catch (error) {
        console.log("Downloading " + filename);
        const res = util.downloadFile(filename);
        const collection = $app.findCollectionByNameOrId("apk");
        const record = new Record(collection, {
            "filename": filename,
            "content": res
        });
        $app.saveNoValidate(record);
        console.log("record saved");
        return util.blob(e, 200, "application/octet-stream", res);
    }
});
```

And util.js (not really relevant, but for completeness).
```
module.exports = {
    downloadFile: (filename) => {
        const res = $http.send({
            url: "https://dl-cdn.alpinelinux.org/alpine/latest-stable/main/x86_64/" + filename,
            method: "get",
        }).raw
        return res;
    },
    blob: (e, code, ct, body) => {
        e.response.header().set("content-type", ct)
        e.response.writeHeader(code)
        e.response.write(body)
        return null // optional
    }
}
```

I also have the complete exception of course:
```
[PANIC RECOVER] template: err:376: bad character U+FFFD '�' goroutine 64 [running]:
github.com/pocketbase/pocketbase/apis.NewRouter.panicRecover.func3.1()
	/home/runner/work/pocketbase/pocketbase/apis/middlewares.go:269 +0x107
panic({0x15efa00?, 0xc00104e7c0?})
	/opt/hostedtoolcache/go/1.23.3/x64/src/runtime/panic.go:785 +0x132
github.com/dop251/goja.(*Runtime).RunProgram.func1()
	/home/runner/go/pkg/mod/github.com/dop251/goja@v0.0.0-20241009100908-5f46f2705ca3/runtime.go:1450 +0x28d
panic({0x15efa00?, 0xc00104e7c0?})
	/opt/hostedtoolcache/go/1.23.3/x64/src/runtime/panic.go:785 +0x132
github.com/dop251/goja.(*vm).handleThrow(0xc0001cfc20, {0x15efa00, 0xc00104e7c0})
	/home/runner/go/pkg/mod/github.com/dop251/goja@v0.0.0-20241009100908-5f46f2705ca3/vm.go:834 +0x405
github.com/dop251/goja.(*vm).runTryInner.func1()
	/home/runner/go/pkg/mod/github.com/dop251/goja@v0.0.0-20241009100908-5f46f2705ca3/vm.go:876 +0x3f
panic({0x15efa00?, 0xc00104e7c0?})
	/opt/hostedtoolcache/go/1.23.3/x64/src/runtime/panic.go:785 +0x132
github.com/dop251/goja.(*vm).handleThrow(0xc0001cfc20, {0x15efa00, 0xc00104e7c0})
	/home/runner/go/pkg/mod/github.com/dop251/goja@v0.0.0-20241009100908-5f46f2705ca3/vm.go:834 +0x405
github.com/dop251/goja.(*vm).runTryInner.func1()
	/home/runner/go/pkg/mod/github.com/dop251/goja@v0.0.0-20241009100908-5f46f2705ca3/vm.go:876 +0x3f
panic({0x15efa00?, 0xc00104e7c0?})
	/opt/hostedtoolcache/go/1.23.3/x64/src/runtime/panic.go:785 +0x132
text/template.Must(...)
	/opt/hostedtoolcache/go/1.23.3/x64/src/text/template/helper.go:26
github.com/go-ozzo/ozzo-validation/v4.ErrorObject.Error({{0x1806f61, 0x17}, {0xc001818000, 0x737f1}, 0xc000a23410})
	/home/runner/go/pkg/mod/github.com/go-ozzo/ozzo-validation/v4@v4.3.0/error.go:110 +0x1df
github.com/go-ozzo/ozzo-validation/v4.Errors.Error(0xc000a23110)
	/home/runner/go/pkg/mod/github.com/go-ozzo/ozzo-validation/v4@v4.3.0/error.go:137 +0x39d
github.com/dop251/goja.(*Runtime).NewGoError(0xc000244008, {0x1f08fa0, 0xc000a23110})
	/home/runner/go/pkg/mod/github.com/dop251/goja@v0.0.0-20241009100908-5f46f2705ca3/runtime.go:549 +0x4a
github.c
```

**Reply by ganigeorgiev** [reply]
It seems to be triggered by the error serialization. What is the collection field type of the `content` field?

**Reply by ganigeorgiev** [reply]
If `content` is of type `file` then storing the raw bytes as its value is wrong (that would explain why the validation fails). The expected value of a `file` type is string (or array of strings) representing the uploaded file name(s).

Previously it could have "worked" because most likely you inserted the data with `dao().saveRecord()` which doesn't perform validation (aka. is is roughly equivalent to a raw INSERT/UPDATE).

Could you elaborate what exactly is the use case here?

**Reply by petrm** [reply]
`content` is of type `file` with no mime type set (I don't seem to be able to select N/A octet-type from the dropdown).
```
[
    {
        "id": "nhij3hyy3kk6jwa",
        "listRule": "",
        "viewRule": null,
        "createRule": null,
        "updateRule": null,
        "deleteRule": null,
        "name": "apk",
        "type": "base",
        "fields": [
            {
                "autogeneratePattern": "[a-z0-9]{15}",
                "hidden": false,
                "id": "text3208210256",
                "max": 15,
                "min": 15,
                "name": "id",
                "pattern": "^[a-z0-9]+$",
                "presentable": false,
                "primaryKey": true,
                "required": true,
                "system": true,
                "type": "text"
            },
            {
                "hidden": false,
                "id": "lbn5cyiv",
                "maxSelect": 1,
                "maxSize": 5242880,
                "mimeTypes": [],
                "name": "content",
                "presentable": false,
                "protected": false,
                "required": true,
                "system": false,
                "thumbs": null,
                "type": "file"
            },
            {
                "autogeneratePattern": "",
                "hidden": false,
                "id": "vugvdool",
                "max": 0,
                "min": 0,
                "name": "filename",
                "pattern": "",
                "presentable": false,
                "primaryKey": false,
                "required": true,
                "system": false,
                "type": "text"
            },
            {
                "hidden": false,
                "id": "autodate2990389176",
                "name": "created",
                "onCreate": true,
                "onUpdate": false,
                "presentable": false,
                "system": false,
                "type": "autodate"
            },
            {
                "hidden": false,
                "id": "autodate3332085495",
                "name": "updated",
                "onCreate": true,
                "onUpdate": true,
                "presentable": false,
                "system": false,
                "type": "autodate"
            }
        ],
        "indexes": [
            "CREATE UNIQUE INDEX `idx_9jCOKIu` ON `apk` (`filename`)"
        ],
        "system": false
    }
]
```

**Reply by petrm** [reply]
As for the use case, I am essentially making a cache for APK files for booting a machine over iPXE. The table stores binary package files from upstream.

**Reply by petrm** [reply]
> If `content` is of type `file` then storing the raw bytes as its value is wrong (that would explain why the validation fails). The expected value of a `file` type is string (or array of strings) representing the uploaded file name(s).
> 
> Previously it could have "worked" because most likely you inserted the data with `dao().saveRecord()` which doesn't perform validation (aka. is is roughly equivalent to a raw INSERT/UPDATE).
> 
> Could you elaborate what exactly is the use case here?

You are correct. I used  `dao().saveRecord()` before and migrated this to `$app.save(record)` as the docs recommended.

**Reply by ganigeorgiev** [reply]
OK, there seems to be a misunderstanding here.

Again, the `file` type field is intended to store only the file name(s) as reference to filesytem storage (local or S3) and it is not intended for storing blobs. I'm actually suprised that this previously even worked but it is most likely because the validations were previously executed only if you were updating the record with the `RecordUpsert` form, which is superseded with just `$app.save()` in v0.23.0.

Please elaborate what is the end goal here. If you want to upload files, you can refer to the example in https://pocketbase.io/docs/js-records/#create-new-record.

For serving back that file you may not need a custom action and can use the buitin [`/api/files/*`](https://pocketbase.io/docs/files-handling/#file-url) endpoint.

Alternatively, if you really want to serve the uploaded record file manually, then you can check https://pocketbase.io/docs/js-filesystem/#reading-files. To serve the file you can also pass the reader directly to [`e.stream()`](https://pocketbase.io/jsvm/interfaces/core.RequestEvent.html#stream).

**Reply by ganigeorgiev** [reply]
I'll submit a fix for the panic to have more user friendly error but again the above usage is wrong.

**Reply by petrm** [reply]
Thanks for the support Gani, super helpful.
I need to serve the files from a specific URL, so the built-in endpoint doesn't work for me. Some of files are also generated on-the-fly. They are also not uploaded by the client, they are fetched by pocketbase and stored in that table.

I used the `file` type when creating the collection, as I didn't see any other suitable field type for binary content. Sqlite doesn't really care what is stuffed in the column, but I don't think using string is a good idea either. Is there another field type that I am missing?

**Reply by petrm** [reply]
Again many thaks for pointing me to the filesystem abstraction. I found other methods that simplify my code.

However, I think there is some inconsistency in the docs:
https://pocketbase.io/docs/js-filesystem/#saving-files refers to `$filesystem.NewFileFromURL(url)`
Then there is:
https://pocketbase.io/jsvm/interfaces/filesystem.newFileFromURL.html - please note that it starts with small letter.
And neither one can be found by my app:
`TypeError: Object has no member 'newFileFromURL' at <eval>:4:44(19)` and `TypeError: Object has no member 'NewfileFromURL' at <eval>:5:44(26)`

Then I found `https://pocketbase.io/jsvm/functions/_filesystem.fileFromUrl.html` which works as expected.

What am I missing here?

**Reply by ganigeorgiev** [reply]
A fix for the panic with a more friendly error message was added in [v0.23.1](https://github.com/pocketbase/pocketbase/releases/tag/v0.23.1).

Regarding the `$filesystem.*` methods - the example in the website documentation is wrong it must start with lower case (I'll update it a little bit later today).

The example in autogenerated namespace docs in the JSVM reference docs unfortunately  cannot be updated easily because it is generated from the Go source code.

**Reply by ganigeorgiev** [reply]
For the `$filesystem.*` JSVM reference you can check https://pocketbase.io/jsvm/modules/_filesystem.html.

It is a little confusing but note that the `$filesystem.*` under the hood utilizes the Go generated namespace type `filesystem` but the latter is just a type interface and not an actual binding.

**Reply by ganigeorgiev** [reply]
Since this discussion started to become a little messy, here is an alternative version of your custom route using the filesystem storage (_treat it as a pseudo-code as I haven't tested it_):

```js
routerAdd("GET", "/boot/x86_64/{filename}", (e) => {
    const filename = e.request.pathValue("filename");
    const collection = $app.findCollectionByNameOrId("apk");

    // retrieve/create a record reference for the file
    let record;
    try {
        record = $app.findFirstRecordByData(collection, "filename", filename);
        console.log("found file " + filename);
    } catch (error) {
        // maybe filter/sanitize the filename?
        console.log("Downloading " + filename);
        let file = $filesystem.fileFromUrl("https://dl-cdn.alpinelinux.org/alpine/latest-stable/main/x86_64/" + filename)
        record = new Record(collection);
        record.set("filename", filename)
        record.set("content", file)
        $app.save(record);
        console.log("record saved");
    }

    // serve the file
    let fsys, reader;
    try {
        fsys = $app.newFilesystem();
        reader = fsys.getFile(record.baseFilesPath() + "/" + record.get("content"))

        e.stream(200, "application/octet-stream", reader)
    } finally {
        reader?.close();
        fsys?.close();
    }
});
```

If you want to keep your old version, aka. storing the files in the database as blobs then this officially is not support and the closest type for storing raw bytes would be the `json` field type.
